### PR TITLE
webextension permission parsing

### PR DESCRIPTION
**Pull request by eviljeff**
First part of #3981 - does the extracting.
Permission descriptions (and pretty names) and defined as constants here because it's the only easy way to get them localised.  Long term a better solution should be swapped in (i'll file some issues)

### CODE REVIEW

File: src/olympia/files/models.py [hunk 1]
```diff
@@ -706,6 +712,50 @@ class Meta:
         db_table = 'validation_annotations'
 
 
+class WebextPermission(ModelBase):
```
Comment (diox) [inline src/olympia/files/models.py]: It feels over-engineered to have a whole model for each permission, and then a number of many2many rows for each add-on. Do we actually need that granularity ? It's going to cost us each time we want to display the list. I would have just stored a `JSONField` with the permission names per `File` and that's it (either in `File` directly or in a separate model with a `OneToOneField` to `File`) ...

File: src/olympia/files/models.py [hunk 2]
```diff
@@ -706,6 +712,50 @@ class Meta:
...
+    KNOWN_PERMISSIONS = WEBEXT_PERMISSIONS_DICT
+    name = models.CharField(max_length=255, unique=True, null=False)
+    _enum = models.IntegerField(default=WEBEXT_UNKNOWN.id, db_column='enum')
```
Comment (diox) [inline src/olympia/files/models.py]: Assuming we do keep the `ManyToManyField` approach (see above) I'm not a fan of keeping yet another set of ids around. We could just use the name to find the description + set some `choices` on the `name` field, which would give us `get_name_display()` for free.

File: src/olympia/files/models.py [hunk 3]
```diff
@@ -706,6 +712,50 @@ class Meta:
...
+    files = models.ManyToManyField(File, related_name='webext_permissions',
+                                   through='FileWebextPermission')
```
Comment (diox) [inline src/olympia/files/models.py]: nit: Again, not a fan of the `ManyToManyField` approach but I'd make more sense to have this on the `File` class if we do that

File: src/olympia/files/models.py [hunk 4]
```diff
@@ -706,6 +712,50 @@ class Meta:
...
+
+    class Meta(ModelBase.Meta):
+        db_table = 'webext_permissions'
+
+    def __init__(self, *args, **kw):
+        super(WebextPermission, self).__init__(*args, **kw)
+
+        enum = WEBEXT_PERMISSIONS_DICT.get(self.name, WEBEXT_UNKNOWN).id
+        self.update(_enum=enum)
```
Comment (diox) [inline src/olympia/files/models.py]: We don't want to trigger a database save each time we instantiate this class... Another strike against having `enum` for me

File: src/olympia/files/models.py (hunk 2 above)
Comment (diox) [inline src/olympia/files/models.py]: Bonus: we wouldn't even need object constants (that are not terribly useful) with this idea.

File: src/olympia/files/models.py (hunk 1 above)
Comment (eviljeff) [inline src/olympia/files/models.py]: JSON in a database field is ... icky. Plus so much unnecessary duplication (admittedly relative to the amount of data already in the database, minimal).  Unless there's a performance hit I lean towards ~~over~~correct-engineering.

File: src/olympia/files/models.py (hunk 2 above)
Comment (eviljeff) [inline src/olympia/files/models.py]: fair point - I started with the constants and enum before I'd finished the model.  
We can't do choices alas - there will be permissions we don't yet know about (e.g. a Chrome one; a new one that's just landed) and also urls for match patterns that I think we should keep.

File: src/olympia/files/models.py (hunk 3 above)
Comment (eviljeff) [inline src/olympia/files/models.py]: Okay.

File: src/olympia/files/models.py (hunk 4 above)
Comment (eviljeff) [inline src/olympia/files/models.py]: okay, okay, you've sold me on dropping enum 🙄

**Comment by andymckay** [comment]
As an aside, Firefox doesn't support this yet, but a permission in Chrome could be an object, for example: https://developer.chrome.com/apps/fileSystem and not a simple string. Secondly, how would this deal with hosts, would it ignore them?

**Comment by eviljeff** [comment]
I was planning on keeping hosts in the database after we'd parsed them, though I'm not sure what we'd do with them exactly.  Depends what UX/Product want to show to the user I suppose.

As for objects... wasn't aware of that, thanks!  We'll need another field at least.  Isn't https://developer.chrome.com/extensions/declare_permissions the full list then?  Is fileSystem the only exception?

**Comment by andymckay** [comment]
Those are the extension permissions, the filesystem permissions are app based permissions which we don't support in Firefox. But... we could or do something similar. I just don't want you to get locked into assuming its always going to be a list of strings.

**Comment by andymckay** [comment]
Ignore my last comment, looks like Firefox schema assumes they will always be a list of strings.

File: src/olympia/files/models.py (hunk 1 above)
Comment (diox) [inline src/olympia/files/models.py]: well `ManyToManyField` isn't great for performance to start with... plus the cost of instantiating the model instances themselves on the python side... and the fact that there can be arbitrary strings for the hosts... all this really pushes me towards storing a simple dumb list.

The only argument I can see for being more granular is that it provides the ability to match files using a given permission easily, but that's also going to be possible with a (true) json field: https://dev.mysql.com/doc/refman/5.7/en/json.html#json-paths , just not from the Django ORM right now.

**Comment by EnTeQuAk** [comment]
Stupid side question here: We are going to pull the schema.json from mozilla-central at some point, does someone know where the permission descriptions are stored in the Firefox repository? It's a long shot but we start re-defining this again, similar to app-versions and I feel like this could be yet another maintenance burden I would love to avoid if possible.

**Comment by eviljeff** [comment]
https://github.com/mozilla/addons-server/pull/4519/commits/913e2b07204b1e4979f43f7b0110506dd6e04ef0 now with less engineering 🔨

File: src/olympia/files/models.py [hunk 5]
```diff
@@ -82,6 +83,8 @@ class File(OnChangeMixin, ModelBase):
     # STATUS_NULL means the user didn't disable the File - i.e. Mozilla did.
     original_status = models.PositiveSmallIntegerField(
         default=amo.STATUS_NULL)
+    webext_permissions_json = models.CharField(
+        max_length=1024, default='')
```
Comment (diox) [inline src/olympia/files/models.py]: As I mentioned before, you could use a nullable `OneToOneField` + a separate model. In any case, you want a `JSONField` (from `django_extensions.db.fields.json`) - and probably no `max_length`

File: src/olympia/files/models.py [hunk 6]
```diff
@@ -176,6 +179,10 @@ def from_upload(cls, upload, version, platform, is_beta=False,
             if validation['metadata'].get('requires_chrome'):
                 file_.requires_chrome = True
 
+        if file_.is_webextension and parsed_data.get('permissions'):
+            file_.webext_permissions_json = json.dumps(
```
Comment (diox) [inline src/olympia/files/models.py]: With a `JSONField` the `json.dumps()` would be done behind the scenes for you.

File: src/olympia/files/models.py [hunk 7]
```diff
@@ -372,6 +379,17 @@ def get_localepicker(self):
         statsd.timing('files.extract.localepicker', (end * 1000))
         return res
 
+    @amo.cached_property
+    def webext_permissions(self):
+        return {name: WEBEXT_PERMISSIONS.get(name, Permission(name, '', ''))
+                for name in json.loads(self.webext_permissions_json)}
```
Comment (diox) [inline src/olympia/files/models.py]: Similarly, with a `JSONField` the `json.loads()` would be done behind the scenes for you.

**Comment by andymckay** [comment]
To answer @EnTeQuAk its here: https://dxr.mozilla.org/mozilla-central/source/browser/locales/en-US/chrome/browser/browser.properties#36-120, for more questions aswan is the expert.

**Comment by eviljeff** [comment]
https://github.com/mozilla/addons-server/pull/4519/commits/ce881f5ca29fac729b418fb23bb9a803f73ee90f with separate model.  Also flattened permission constants into a dict of tuples.

File: src/olympia/files/models.py [hunk 8]
```diff
@@ -706,6 +724,12 @@ class Meta:
         db_table = 'validation_annotations'
 
 
+class WebextPermission(ModelBase):
+    permissions = JSONField(default={})
+    file = models.ForeignKey('File', related_name='_webext_permissions',
```
Comment (diox) [inline src/olympia/files/models.py]: I'd make this a `OneToOneField` and adjust the code above to match. We should never have a `File` with multiple `WebextPermission`s

Comment (eviljeff) [inline src/olympia/files/models.py]: https://github.com/mozilla/addons-server/pull/4519/commits/e113883304ccb8a9db1c14db351e98d89be9c2db

**Pull request merged**